Add smart indentation for brace completion.

NTaylorMullen opened the pull request:
This is a middle-ground approach to implementing brace completion smart indent without refactoring the entire `VisualStudioRazorParser`. Decided to use the "smart indent" verbiage because that's what the visual studio editor uses.

Here's what this PR accomplishes. Pipe | indicates cursor

## Razor code block smart indentation
A user types `@{|`, Roslyn autocompletes it to `@{ |}`. User then hits enter and **without** this PR it becomes:
```
@{
    |}
```
This PR fixes that and makes it:
```
@{
    |
}
```

## Razor directive block smart indentation
A user types `@functions {|`, Roslyn autocompletes it to `@functions {|}`. User then hits enter and **without** this PR it becomes:
```
@functions {
    |}
```
This PR fixes that and makes it:
```
@functions {
    |
}
```

One significant note is that the cursor position in the above examples don't have spaces before it. That's the editor auto-indent that will auto-add those spaces if a user then attempts to add content.

---------------------------------------------------------------------------------

- Added a standalone brace smart indenter that listens to `ITextBuffer` changed events to determine when a brace completion event needs to be handled.
- Added extension methods to handle interaction with the `ITextBuffer` and document trackers and text views.
- Added a `BraceSmartIndenterTest` to verify all parts of the smart indenter.
- Moved private test infrastructure classes into their own files and expanded on their functionality to enable the brace completion smart indent scenarios.

#1538

NTaylorMullen commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,218 @@
+    /// }
+    /// This is also true for directive block scenarios.
+    /// </example>
+    internal class BraceSmartIndenter : IDisposable
```
`IDisposable` for now since the `VisualStudioRazorParser` is. This will probably change once we own the lifetime of the parser (my next item).

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,218 @@
+                var currentSnapshotPoint = new SnapshotPoint(textView.TextBuffer.CurrentSnapshot, caret.Position);
+                textView.Caret.MoveTo(currentSnapshotPoint);
+
+                // For Razor metacode cases the editors smart indent wont kick in automatically. This forces the smart indent.
```
`editor's`

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,218 @@
+
+                // For Razor metacode cases the editors smart indent wont kick in automatically. This forces the smart indent.
+                // For example attempting to do enter a newline between the functions directive: @functions {} will not auto-indent
+                // inbetween the braces unless we forcefully move to end of line.
```
`in between`

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,218 @@
+            var lastChange = contentChange.Changes[contentChange.Changes.Count - 1];
+            var oldLength = lastChange.OldEnd - firstChange.OldPosition;
+            var newLength = lastChange.NewEnd - firstChange.NewPosition;
+            var newText = contentChange.After.GetText(firstChange.NewPosition, newLength);
```
Can we put this in a private method called `GetNormalizedChange` - if this is duplicated with the parser maybe an extension method?

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,218 @@
+            var focusedTextView = textBuffer.GetFocusedTextView();
+            if (focusedTextView != null && ParserHelpers.IsNewLine(finalText))
+            {
+                var currentSnapshot = textBuffer.CurrentSnapshot;
```
I don't feel like I super understand what this code is doing

NTaylorMullen replied:
I'll pull pieces out into methods to be more descriptive.

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/TextBufferExtensions.cs:
```diff
@@ -17,5 +20,34 @@ public static bool IsRazorBuffer(this ITextBuffer textBuffer)
             return textBuffer.ContentType.IsOfType(RazorLanguage.ContentType);
         }
+
+        public static VisualStudioDocumentTracker GetDocumentTracker(this ITextBuffer textBuffer)
```
Why do we need this? Can you just use the service?

rynowak replied:
I really really don't like the idea of hanging lots of interdependent functionality on extensions, I'd prefer that it use services.

For instance you could add `GetFocusedTextView` to the document tracker service

NTaylorMullen replied:
Sure we can move these

rynowak commented on test/Microsoft.VisualStudio.Editor.Razor.Test.Common/StringTextSnapshot.cs:
```diff
@@ -4,15 +4,32 @@
+            while (delimiterIndex != -1)
+            {
+                delimiterIndex = Content.IndexOf(Environment.NewLine, start);
+                var nextLineStartIndex = delimiterIndex != -1 ? delimiterIndex + Environment.NewLine.Length : Content.Length;
```
This isn't right, we need to be able to work with all line break types in all environements

rynowak commented on test/Microsoft.VisualStudio.Editor.Razor.Test/Infrastructure/TestEdit.cs:
```diff
@@ -0,0 +1,31 @@
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.Editor.Razor.Test.Infrastructure
```
Don't use gross namespaces, just put this in `Microsoft.VisualStudio.Test`

rynowak commented on test/Microsoft.VisualStudio.Editor.Razor.Test/BraceSmartIndenterTest.cs:
```diff
@@ -0,0 +1,382 @@
+            var editorOperationsFactory = new Mock<IEditorOperationsFactoryService>();
+            editorOperationsFactory.Setup(factory => factory.GetEditorOperations(It.IsAny<ITextView>()))
+                .Returns(editorOperations.Object);
+            var braceSmartIndenter = new BraceSmartIndenter(Dispatcher, textBuffer, editorOperationsFactory.Object);
```
This is really opaque and hard to understand, can we move some of this boilerplate to properties

NTaylorMullen replied:
Why not have it opaque? It's a standalone component.

rynowak commented on test/Microsoft.VisualStudio.Editor.Razor.Test/BraceSmartIndenterTest.cs:
```diff
@@ -0,0 +1,382 @@
+        public void TextBuffer_OnPostChanged_MovesToEndOfLine()
+        {
+            // Arrange
+            var change = Environment.NewLine;
```
This isn't right, we need to work with all newline types in all places, we shouldn't test based on `Environment.NewLine`

NTaylorMullen replied:
Ya, getting all the test infrastructure to work with all variants of newline is painful which is why I stuck to `Environment.NewLine`. The product code on the other hand does understand all aspects. I'll put in some time to make the test infrastructure more resilient.

rynowak commented on test/Microsoft.VisualStudio.Editor.Razor.Test/BraceSmartIndenterTest.cs:
```diff
@@ -0,0 +1,382 @@
+        }
+
+        [Fact]
+        public void TextBuffer_OnPostChanged_PlacesCursorBackToOriginalLocation()
```
`TextBuffer_OnPostChanged_MovesToEndOfLine` and `TextBuffer_OnPostChanged_PlacesCursorBackToOriginalLocation` - what's different about these cases?

NTaylorMullen replied:
They're exactly the same just verifying two different aspects of what should happen. In editor land those aspects have impacts on what the user sees. Would rather have 1 test for each so if we break one of those expectations down the line we have 1 specific test failing rather than 1 multi-purpose test.

rynowak replied:
> They're exactly the same just verifying two different aspects of what should happen. 

But why though

> Would rather have 1 test for each so if we break one of those expectations down the line we have 1 specific test failing rather than 1 multi-purpose test.

This just proves that the factoring is wrong. 

These tests (as written) are integration tests. Unit tests have the goal that you're stating, but these are not that.

NTaylorMullen commented:
🆙 📅 

Addressed all feedback. Key changes are:
- Added a BraceSmartIndenter integration test class (removed the move to end of line and caret tests)
- Refactored the smart indenter class to be more unit testable. This also involved adding a common base class to share with the integration test variant.
- Updated VisualStudioRazorParser to consume document trackers only on foreground threads. This was required when we added GetTracker to the document tracker services for ITextBuffer
- Moved text buffer change logic into an extension method.

NTaylorMullen commented on src/Microsoft.VisualStudio.Editor.Razor/TextContentChangedEventArgsExtensions.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    internal static class TextContentChangedEventArgsExtensions
+    {
+        public static bool TextChangeOccurred(this TextContentChangedEventArgs args, out (ITextChange firstChange, ITextChange lastChange, string newText, string oldText) changeInformation)
```
Add tests.

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,266 @@
+                //     |
+                // }
+            }
+        }
```
This is solid 👍

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,266 @@
+
+            context = null;
+            return false;
+        }
```
This makes way more sense now 👍

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/VisualStudioRazorParser.cs:
```diff
@@ -79,6 +108,8 @@ public VisualStudioRazorParser(ForegroundDispatcher dispatcher, ITextBuffer buff
 
         public string FilePath { get; }
 
+        private readonly ErrorReporter _errorReporter;
```
put this with the fields yo

NTaylorMullen replied:
#thanksvs

rynowak commented on src/Microsoft.VisualStudio.LanguageServices.Razor/Editor/DefaultVisualStudioDocumentTrackerFactory.cs:
```diff
@@ -114,6 +114,25 @@ public override VisualStudioDocumentTracker GetTracker(ITextView textView)
+            DefaultVisualStudioDocumentTracker tracker;
+            if (!textBuffer.Properties.TryGetProperty(typeof(VisualStudioDocumentTracker), out tracker))
+            {
+                Debug.Fail("The document tracker should be initialized");
```
This isn't right, what if the buffer isn't a razor buffer? This should just return null gracefully.

The ITextView version has the asserts because at that point we've already determined that it's a razor buffer

NTaylorMullen replied:
Added a RazorBuffer check above and added tests.

rynowak commented on test/Microsoft.VisualStudio.Editor.Razor.Test/BraceSmartIndenterIntegrationTest.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    public partial class BraceSmartIndenterIntegrationTest : BraceSmartIndenterTestBase
```
Which is it? are they using inheritance or partials?

NTaylorMullen replied:
lmao, well lets make this a `public partial abstract internal protected facepalm class`

rynowak commented on test/Microsoft.VisualStudio.Editor.Razor.Test/Infrastructure/TestTextBuffer.cs:
```diff
@@ -0,0 +1,149 @@
+
+namespace Microsoft.VisualStudio.Test
+{
+    public class TestTextBuffer : ITextBuffer
```
Did you ask about those mocks?

NTaylorMullen replied:
Talked to the WTE folks and they rely on a privately constructed MEF service catalog (probing paths etc.) to create buffer factories; insufficient for our usecase. Will need to reach out to others.

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,266 @@
+            while (linePosition < lineSnapshot.Length)
+            {
+                if (!char.IsWhiteSpace(lineText[linePosition]))
+                    break;
```
braces

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,266 @@
+            while (linePosition >= 0)
+            {
+                if (!char.IsWhiteSpace(lineText[linePosition]))
+                    break;
```
braces

rynowak commented on src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs:
```diff
@@ -0,0 +1,266 @@
+                if (!char.IsWhiteSpace(lineText[linePosition]))
+                    break;
+                linePosition--;
+            }
```
This loop might be simpler written as a `for`

NTaylorMullen replied:
#taylorcontrolflow

NTaylorMullen commented:
🆙 📅
